source/common/tracing/opentracing_driver_impl.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/tracing/http_tracer.h"
+
+#include "common/common/logger.h"
+
+#include "opentracing/tracer.h"
+
+namespace Envoy {
+namespace Tracing {
+
+class OpenTracingSpan : public Span, Logger::Loggable<Logger::Id::tracing> {
+public:
+  OpenTracingSpan(bool use_single_header_propagation, bool use_tracer_propagation,
+                  std::unique_ptr<opentracing::Span>&& span);
+
+  // Tracing::Span
+  void finishSpan() override;
+  void setOperation(const std::string& operation) override;
+  void setTag(const std::string& name, const std::string& value) override;
+  void injectContext(Http::HeaderMap& request_headers) override;
+  SpanPtr spawnChild(const Config& config, const std::string& name, SystemTime start_time) override;
+
+private:
+  bool use_single_header_propagation_;
+  bool use_tracer_propagation_;
+  std::unique_ptr<opentracing::Span> span_;
```

Comment:
is there any way to avoid a double allocation for each span? I haven't looked at all at the OT C++ code yet.

Reply:
It currently also allocating twice since the stored [lightstep::Span](https://github.com/envoyproxy/envoy/blob/master/source/common/tracing/lightstep_tracer_impl.h#L44) is just an [std::shared_ptr](https://github.com/lightstep/lightstep-tracer-cpp/blob/v0_38/src/c%2B%2B11/lightstep/span.h#L54).

I think the double allocation comes from having two levels of abstraction (both at the OT level and within envoy).

It could be eliminated by redesigning Envoy's Span class to be a lightweight stack allocated object that manages a pointer to either an opentracing span or a zipkin span and implements dynamic dispatch via a conditional instead of a vtable. But I don't know how elegant that is.